src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +30,155 @@ pub trait SystemTimeSource: Send + Sync {
 #[dill::interface(dyn SystemTimeSource)]
 pub struct SystemTimeSourceDefault;
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceDefault {
     fn now(&self) -> DateTime<Utc> {
         Utc::now()
     }
+
+    async fn sleep(&self, duration: Duration) {
+        let std_duration = duration.to_std().unwrap();
+
+        tokio::time::sleep(std_duration).await;
+    }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+#[derive(Debug)]
+struct AwaitingCaller {
+    wake_up_time: DateTime<Utc>,
+    waker_tx: Option<oneshot::Sender<()>>,
```

Comment:
I think you earlier had a sequence ID. It might be useful to define some stricter order for priority queue. Time first, then the ID. As `waker_tx` is unlikely usable for sorting.

Reply:
Done